core/common/src/main/java/alluxio/util/executor/SerializedTaskRunner.java

```diff
@@ -0,0 +1,138 @@
+  private volatile boolean mIsShutdown;
+  private final LinkedBlockingQueue<Runnable> mTasks;
+  private volatile Thread mRunningThread;
+
+  /**
+   * Create a new instance of {@link SerializedTaskRunner}.
+   */
+  public SerializedTaskRunner() {
+    mTasks = new LinkedBlockingQueue<>();
+    mRunning = true;
+    mIsShutdown = false;
+  }
+
+  /**
+   * Add a new task to execute into the queue.
+   *
+   * @param r task to run
+   *
+   * @return true if the task was added to the queue successfully
+   */
+  public boolean addTask(Runnable r) {
+    if (mIsShutdown) {
+      return false;
+    }
+    return mTasks.add(r);
+  }
+
+  @Override
+  public void run() {
+    mRunningThread = Thread.currentThread();
+    while (mRunning) {
+      try {
+        Runnable task = mTasks.take();
+        try {
+          task.run();
+        } catch (RuntimeException e) {
+          LOG.warn("Exception in serialized task runner: ", e);
+        }
+      } catch (InterruptedException e) {
+        LOG.debug("Interrupted while waiting for task");
+      }
+    }
+  }
+
+  /**
+   * Shuts down the task runner.
+   *
+   * No new tasks are accepted after this method is called. Remaining tasks are attempted to all
+   * be executed within the shutdown timeout. Otherwise, after the timeout the running thread
+   * will exit, any remaining tasks will fail to ever be executed.
+   *
+   * @param time the time to spend before aborting any outstanding tasks
+   * @param unit the {@link TimeUnit} corresponding to the time argument
+   */
+  public synchronized void shutdown(long time, TimeUnit unit) {
+    Preconditions.checkNotNull(unit, "time unit");
+    if (mIsShutdown) {
+      return;
+    }
+    mIsShutdown = true;
+    long totalTime = unit.toMillis(time);
+    long curr;
+    long start = curr = System.currentTimeMillis();
+    while (mTasks.size() > 0 || curr - start > totalTime) {
+      curr = System.currentTimeMillis();
+      Thread.yield();
+    }
+    mRunning = false;
+    mRunningThread.interrupt();
+    mRunningThread = null;
+  }
+
+  /**
+   * A Factory which can spawn new {@link SerializedTaskRunner}s on separate threads using an
+   * available {@link ExecutorService}.
+   */
+  public static class Factory {
+    private final ExecutorService mService;
+
+    /**
+     * Create a new instance of  {@link Factory}.
+     *
+     * @param svc the executor service backing this factory
+     */
+    public Factory(ExecutorService svc) {
+      mService = svc;
+    }
+
+    /**
+     * Spawn a new task executor which can begin queueing new tasks.
+     *
+     * It is not guaranteed to immediately begin running tasks. It will depend on the underlying
+     * executor's ability to allocate a thread to this task runner.
+     *
+     * @return a new {@link SerializedTaskRunner}
+     */
+    public SerializedTaskRunner create() {
+      SerializedTaskRunner runner = new SerializedTaskRunner();
+      mService.execute(runner);
```

Comment:
Does this mean every new `SerializedTaskRunner` will hold a thread off this pool?

Reply:
Yes, until the task runner is shut down

Reply:
Is this the same with gRPC's `SerializingExecutor` ?

Reply:
As far as I'm aware, no

Reply:
Then maybe we should switch back to it for now?

Reply:
updates